### DISCUSSION TITLE
gp tracer

### DISCUSSION THREAD

**Discussion by Ulf3000** [chsh2/nijiGPen#36] [Ideas]
check this out. (youve probably seen it already ) 
https://www.blendernation.com/2023/01/06/unicorn-wars-made-with-blender-animated-feature-now-in-french-cinemas/

the addon "gp tracer" converts 3d geometry to gp, so it can be edited and finetuned in 2d. i know you are the only dude who can make something similar. Its doesnt need to go through cycles though. i bet you already have a few ideas on how this can be done even easier/faster. I think some of the needed code is already in your nijigp

**Comment by chsh2** [comment]
I am currently working on generating toon shading for Grease Pencil strokes (without having to convert them to meshes first). Maybe it can achieve similar effects as what you posted, but I am not sure yet. I think I still need much time to investigate. I will update once having some progresses.

By the way, how about asking the author of GP Tracer whether they plan to publish it? They have published several popular Grease Pencil add-ons:
https://github.com/Pullusb

**Reply by Ulf3000** [reply]
it basically flattens a 3d model with shading and lines to a greasepencil object with the same size in camera view. means you can animate a 3d model and convert it to grease pencil and then do the fixes and overpainting in GP. They used it in that movie to translate the horses from 3d to 2d becasue they were so hard to animate otherwise. They render the horses with freestyle lines to an image sequnce and then use a vectorizer library to convert it to vectors and import those again as grease pencil frames.

it could be immensly usefull for basically everything, from turning spaceshuttles to fire animation.

i didnt think theyll publish it, they have published all the other tools on gitlab, but not that one... maybe because it not universally usable(hardcoded paths and stuf like that) or becasue its really their secret weapon and nothing you just give to competitors for free.
https://gitlab.com/autour-de-minuit/blender

**Reply by chsh2** [reply]
Thank you for the GitLab link. I do not know it before, and the scripts look very useful.

I am not certain if GP Tracer uses vectorization, because in the demo video the add-on can output base colors and shadows to different GPencil layers, and the light sources need to be manually set. It implies that the add-on may have its own lighting model to calculate the color of every vertex of the 3D model. If it is true, that should be a great amount of work.

Using vectorization on render results may be easier. It is just that we cannot separate the base color and shadow colors. All the output strokes will be in the same GPencil layer.

For the line art part, we can use the [Reproject](https://docs.blender.org/manual/en/latest/grease_pencil/modes/edit/grease_pencil_menu.html#reproject) operator to flatten the baked results of the line art modifier. But maybe the scale of the line art is different from the vectorization results. I may need to write some scripts to resize it.

**Comment by b4zz4** [comment]
Exporting "toon shader" with lineart in 3D and Importing a sequence of images and vectoring it with nijiGPen you have almost the same. 
With the add-on gp_toolbox you can manage the deep by compensating the size and distance.

**Comment by chsh2** [comment]
Recently I am too busy to make enough progress, but today I made an attempt to add an option "Fit to Camera" to the image vectorization operator. If you vectorize the rendered result image with this option, the generated Grease Pencil will have the same size as the original objects:
![Frame](https://github.com/chsh2/nijiGPen/assets/110356534/dbb9cdf4-f52e-49c2-ad8b-9f7197313b83)

If creating a Line Art modifier, baking it, and then executing `[Grease Pencil] - [Clean Up] - [Reproject Strokes]`, the line art will also be projected to the same place as the flat colors, so they can combine to achieve a 2D toon look.

<img width="450" alt="with_line_art" src="https://github.com/chsh2/nijiGPen/assets/110356534/9dddc360-55d0-42cc-8328-7930ec21ccd4">

I may continue to write some scripts that further simplify this process. I do not want to copy exactly what other add-on authors have already done, but vectorizing the render result may be useful in general and I will think of some more ways to facilitate it.

**Comment by Ulf3000** [comment]
ohh holy , i didnt see your answer till now ! very awesome. will test tommorrow ❤️

**Comment by chsh2** [comment]
I tried to create a new operator to automate mesh-to-gp process:
https://gist.github.com/chsh2/94b70e69d492a32c20fb569cd1b39dc6

It is not very straightforward to use, so I did not put it inside the add-on. It should be used in the following way with NijiGPen v0.9 enabled:

![steps](https://github.com/chsh2/nijiGPen/assets/110356534/2ed90259-fcdf-4c4d-b2ef-030848ef714a)

1. Copy and execute the codes into Blender's Text Editor to register the operator.
2. Select Grease Pencil as the active object.
3. Make sure that the GP object has a `Line Art` modifier, with the source object and output layer properly set.
4. Make sure that the camera view is enabled.
5. Execute the operator in `[Edit] - [Menu Search]`(F3) panel. (The operator must be executed in a 3D Viewport, so please do not run the operator in Text Editor or Console. It can be run from the VSCode extension though.)

The output looks like:

https://github.com/chsh2/nijiGPen/assets/110356534/c312e242-09c4-4402-ad1a-13cbf4ca0b37

Currently, this script works correctly only when the camera does not move, because I made a bug in the add-on v0.9.0 about multi-frame import. I will try to fix it in the next release.

**Reply by Ulf3000** [reply]
even with animation support ! nice

**Reply by devilslayergs-a11y** [reply]
wow! 
does it work in blender 5.0+?
